### PR TITLE
materialized: use PID file for mutual exclusion

### LINKED ISSUE #7816: materialized allows multiple simultaneous copies open pointed at the same catalog
In terminal 1
```
$ ./materialized --version
materialized v0.9.0 (fc3b56e56)
$ ./materialized -D /tmp/mz
=======================================================================
Thank you for trying Materialize!

We are interested in any and all feedback you have, which may be able
to improve both our software and your queries! Please reach out at:

    Web: https://materialize.com
    GitHub issues: https://github.com/MaterializeInc/materialize/issues
    Email: support@materialize.io
    Twitter: @MaterializeInc
=======================================================================

materialized v0.9.0 (fc3b56e56) listening on 0.0.0.0:6875...
```

In terminal 2
```
$ ./materialized -D /tmp/mz --listen-addr=0.0.0.0:6876
=======================================================================
Thank you for trying Materialize!

We are interested in any and all feedback you have, which may be able
to improve both our software and your queries! Please reach out at:

    Web: https://materialize.com
    GitHub issues: https://github.com/MaterializeInc/materialize/issues
    Email: support@materialize.io
    Twitter: @MaterializeInc
=======================================================================

materialized v0.9.0 (fc3b56e56) listening on 0.0.0.0:6876...
```

@benesch tells me this probably a bug and we might be able to prevent it using https://www.sqlite.org/pragma.html#pragma_locking_mode

### PR DESCRIPTION
**Pull request by benesch**
cc @danhhz to check whether this will work for persistence, and @elindsey for assessing whether we're comfortable taking this C dependency.

----

Add a PID file to protect against multiple copies of materialized
running against the same mzdata directory simultaneous. The PID file
lives at mzdata/materialize.log.

Unfortunately we can't rely on SQLite to provide this mutual exclusion.
SQLite allows you to request exclusive access to a database file, but
our tests require the ability to introspect the SQLite database as
materialized is running.

Handling PID files is notoriously tricky, so this patch brings in some C
code from libbsd that is known to get the tricky bits right.

Fix #7816.

### CODE REVIEW

File: src/pid-file/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+use std::os::unix::fs::PermissionsExt;
+use std::path::Path;
+use std::ptr;
+
+use ore::option::OptionExt;
+
+#[allow(non_camel_case_types)]
+#[repr(C)]
+struct pidfh {
+    private: [u8; 0],
```
Comment (elindsey) [inline src/pid-file/src/lib.rs:52]: Is this like the rust equivalent of flexible array members or some such - a way to say "this is some arbitrary pile of bytes allocated/managed on the other side of the ffi boundary, don't worry about it"?

File: src/pid-file/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+            Ok(())
+        } else {
+            Err(Error::Io(io::Error::last_os_error()))
+        }
+    }
+}
+
+impl Drop for PidFile {
+    fn drop(&mut self) {
+        if self.0 != ptr::null_mut() {
```
Comment (elindsey) [inline src/pid-file/src/lib.rs:133]: I'd usually expect this to call self.remove and throw the error away, does something prevent that?

**Review by elindsey — APPROVED** [review]
C lib looks good to me!

A few rust-isms questions for my own edification, but don't feel the need to humor me if you're busy 🙂

File: src/pid-file/src/lib.rs (hunk 2 above)
Comment (danhhz) [reply src/pid-file/src/lib.rs:133]: Remove takes a `mut self`, which consumes it, while Drop impls only get an `&mut self` borrow. One workaround is to have a private `&mut self` that they both call.

I almost suggested that this might have an issue with Drop getting called at the end of remove, but managed to notice that you've already handled that with the `self.0 != ptr::null_mut()`. May I suggest either a comment or tossing the inside bit of PidFile into an Option, which you `if let Some(...) = self.0.take() {` on remove/drop?

File: src/pid-file/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+            Error::AlreadyRunning { pid } => write!(
+                f,
+                "process already running (PID: {})",
+                pid.display_or("<unknown>")
+            ),
+        }
+    }
+}
+
+impl std::error::Error for Error {}
```
Comment (danhhz) [inline src/pid-file/src/lib.rs:173]: Maybe a simple test (like you have for Service)?

**Review by danhhz — APPROVED** [review]
LGTM! Thanks for doing this!

File: src/pid-file/src/lib.rs (hunk 1 above)
Comment (benesch) [reply src/pid-file/src/lib.rs:52]: Yep, exactly. This SO post has a good explanation of the situation: https://stackoverflow.com/a/38315613/1122351

tl;dr an RFC for a better way was proposed and accepted, but naturally it's stuck in committee: rust-lang/rust#43467.

File: src/pid-file/src/lib.rs (hunk 2 above)
Comment (benesch) [reply src/pid-file/src/lib.rs:135]: Yep, added a comment.

File: src/pid-file/src/lib.rs (hunk 3 above)
Comment (benesch) [reply src/pid-file/src/lib.rs:175]: Done.

**Review by benesch — COMMENTED** [review]
Thanks for the reviews!
